exposure-notification/src/test/java/fi/thl/covid19/exposurenotification/efgs/FederationServiceWithDaoTestIT.java

```diff
@@ -94,9 +99,69 @@ public void downloadKeys() {
                 );
         federationGatewayService.startInbound(LocalDate.now(ZoneOffset.UTC), Optional.empty());
 
-        List<TemporaryExposureKey> dbKeys = dd.getIntervalKeys(IntervalNumber.to24HourInterval(Instant.now()));
+        List<TemporaryExposureKey> dbKeys = diagnosisKeyDao.getIntervalKeys(IntervalNumber.to24HourInterval(Instant.now()));
         assertTrue(keys.size() == dbKeys.size() && dbKeys.containsAll(keys) && keys.containsAll(dbKeys));
-        assertDownloadOperationStateIsCorrect(keys.get(1), 10);
+        operationDao.getAndResolveCrashed(EfgsOperationDirection.INBOUND);
+        federationGatewayService.startInboundRetry(LocalDate.now(ZoneOffset.UTC));
+        assertDownloadOperationStateIsCorrect(10);
+    }
+
+    @Test
+    public void downloadRetry() {
+        List<TemporaryExposureKey> keys = FederationGatewayBatchUtil.transform(FederationGatewayBatchUtil.transform(keyGenerator.someKeys(10)));
+        mockServer.expect(ExpectedCount.once(),
+                requestTo("http://localhost:8080/diagnosiskeys/download/" + FederationGatewayBatchUtil.getDateString(LocalDate.now(ZoneOffset.UTC))))
+                .andExpect(method(HttpMethod.GET))
+                .andRespond(withStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .body("")
+                );
+        mockServer.expect(ExpectedCount.once(),
+                requestTo("http://localhost:8080/diagnosiskeys/download/" + FederationGatewayBatchUtil.getDateString(LocalDate.now(ZoneOffset.UTC))))
+                .andExpect(method(HttpMethod.GET))
+                .andRespond(withStatus(HttpStatus.OK)
+                        .contentType(PROTOBUF_MEDIATYPE)
+                        .headers(getDownloadResponseHeaders())
+                        .body(serialize(transform(keys)))
+                );
+        LocalDate date = LocalDate.now(ZoneOffset.UTC);
+        try {
+            federationGatewayService.startInbound(date, Optional.of(TEST_TAG_NAME));
+        } catch (Exception e) {
```

Comment:
For stronger assertion and clearer intent, just catch the one exception you're expecting

Reply:
Yep, that's a good point. I'm planning to sanitize this test class later, but changed these in this point.